### PR TITLE
Adaptive Load command line client

### PR DESCRIPTION
**Pull request by eric846**
An adaptive load command line executable that drives a Nighthawk Service with a series of benchmarks, varying the RPS or another input variable dynamically.

`bazel build -c opt :nighthawk`

`bazel-bin/nighthawk_adaptive_load_client`

- `--spec-file` _pathname_
  - Pathname of an `AdaptiveLoadSessionSpec` textproto defining the session
  - Required
  - Example:`test/adaptive_load/test_data/valid_session_spec.textproto`
- `--output-file` _pathname_ 
  - Pathname to write the `AdaptiveLoadSessionOutput` textproto session output
  - Required
- `--nighthawk-service-address` _host:port_ 
  - Address of the Nighthawk Service to be driven by the adaptive load controller over gRPC
  - Default `localhost:8443`
- `--use-tls`
  - Use TLS for the gRPC connection when sending commands to the Nighthawk Service
  - Default false (insecure connection)

Note that to run an adaptive load session, you must also have a Nighthawk Service running, e.g.:
`bazel-bin/nighthawk_service &`  (to run at localhost:8443)

Part of #416.

### CODE REVIEW

**Comment by dubious90** [comment]
@wjuan-AFK Please review and assign to me when done.  Also note that we believe the CI failure to be unrelated to these changes.

**Comment by oschaaf** [comment]
> @wjuan-AFK Please review and assign to me when done. Also note that we believe the CI failure to be unrelated to these changes.

@dubious90 Created https://github.com/envoyproxy/nighthawk/pull/620 to address the ci trouble with respect to docker.

File: source/adaptive_load/adaptive_load_client_main.h [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+   * @throw Nighthawk::Client::MalformedArgvException If command line constraints are violated.
+   */
+  AdaptiveLoadClientMain(int argc, const char* const* argv, AdaptiveLoadController& controller,
+                         Envoy::Filesystem::Instance& filesystem);
+  /**
+   * Loads the adaptive load session spec proto from a file, runs an adaptive load session, and
+   * writes the output proto to a file. File paths are taken from class members initialized in the
+   * constructor.
+   *
+   * @return Exit code for this process.
```
Comment (wjuan-AFK) [inline source/adaptive_load/adaptive_load_client_main.h:37]: Should we state that 0 is good or is that implicit?

File: source/adaptive_load/adaptive_load_client_main.cc [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+  const Envoy::Api::IoCallBoolResult open_result =
+      file->open(((1 << Envoy::Filesystem::File::Operation::Write)) |
+                 (1 << (Envoy::Filesystem::File::Operation::Create)));
+  if (!open_result.ok()) {
+    throw Nighthawk::NighthawkException(absl::StrCat("Unable to open output file \"", path,
+                                                     "\": ", open_result.err_->getErrorDetails()));
+  }
+  const Envoy::Api::IoCallSizeResult write_result = file->write(contents);
+  if (!write_result.ok()) {
+    throw Nighthawk::NighthawkException(absl::StrCat("Unable to write output file \"", path,
```
Comment (wjuan-AFK) [inline source/adaptive_load/adaptive_load_client_main.cc:52]: nit: write to output file.

**Review by wjuan-AFK — COMMENTED** [review]
LGTM

File: source/adaptive_load/adaptive_load_client_main.cc [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+  Nighthawk::Utility::parseCommand(cmd, argc, argv);
+
+  nighthawk_service_address_ = nighthawk_service_address.getValue();
+  use_tls_ = use_tls.getValue();
+  spec_filename_ = spec_filename.getValue();
+  output_filename_ = output_filename.getValue();
+}
+
+uint32_t AdaptiveLoadClientMain::Run() {
```
Comment (dubious90) [inline source/adaptive_load/adaptive_load_client_main.cc:100]: can we file a future bug to enable some dependency injection of the channel here?  in the future, people might want to enable other ways of connecting to the nighthawk service besides the two provided here.  but we don't need to prematurely allow for that case until we have a specific usecase.

File: source/adaptive_load/adaptive_load_client_main.h (hunk 1 above)
Comment (dubious90) [reply source/adaptive_load/adaptive_load_client_main.h:37]: This is a standard in linux, so we can rely on that understanding.  https://linuxize.com/post/bash-exit/#exit-status

**Review by dubious90 — APPROVED** [review]

File: source/adaptive_load/adaptive_load_client_main.cc (hunk 2 above)
Comment (eric846) [reply source/adaptive_load/adaptive_load_client_main.cc]: Done.

File: source/adaptive_load/adaptive_load_client_main.cc (hunk 3 above)
Comment (eric846) [reply source/adaptive_load/adaptive_load_client_main.cc:100]: Filed #626.

**Review by dubious90 — APPROVED** [review]

**Pull request merged**